### ISSUE TITLE
The blacklist isn't updating if I remove something.

### ISSUE THREAD

**Issue by Burg-Eelen** [kwaschny/unwanted-twitch#30]
I am currently using Firefox 72.0.1 (64-bit), which is not the current version. I seem to recall having similar problems with an older version on another computer last year, which makes me hesitant to believe the following problem could be solved by a simple update. But here's my problem:

When I manage the blacklist and have it synchronized via a cloud, the list keeps what's blocked. But if I mistakenly click an X on a specific stream or category, if I want to remove them, as soon as I click on Save and it tells me to reload the tab, as soon as the tab is reloaded the streams and categories I want unblocked remain blocked.

In the hopes of trying to quickly solve this issue, I tried having my list not synchronized via whatever outside storage is being used, and all it does when I update or import my list is leave the entire list blank after saving. I also tried uninstalling and reinstalling the plugin, and the same problems persist (whether I synchronize with the "cloud" or not).

I'm not sure if I'm missing something or if I messed up a setting somewhere that is causing this problem to happen.

**Comment by kwaschny** [comment]
I need you to help me debugging this. Let's start with checking for possible warnings and errors the extension returns while attempting to modify/save the blacklist.

To do so, open your Firefox, visit twitch.tv and then press the `F12` key (assuming Windows) to open Firefox' developer window. The window has a tab called `Console` (might depend on your system language). Open this one and keep it open at all times. It will report a bunch of (unrelated) things, so don't panic. The things we are interested in are prefixed with `UTTV` (**U**nwanted**T**witch**TV**). You can filter messages in the section by using the input at the top, just enter `UTTV` there.

Now modify and save the blacklist (or whatever action you are having trouble with). Saving in the separate "management" tab will dispatch a message to the other twitch.tv tab, so you will have to check both consoles for warnings and errors (you have to open the developer window for every new tab, so don't forget that). Finally, please share the messages and data you receive. Note that you might be able to expand information in the console (look out for arrows to click).

Hopefully this will already give me some insight.

**Comment by Burg-Eelen** [comment]
My apologies for not getting back to you earlier; I was distracted with other issues.

Thanks for trying to help me out on this, but I'm beginning to think the problem may not be with the program after all, but some other things on my computer. Nevertheless, unless I messed up the instructions, here's the UTTV-related info that showed up on console:

```
3 hidden
UTTV WARN: No sidebar items found on the current page. Provided nodes: 
NodeList []
​
length: 0
​
<prototype>: NodeListPrototype { item: item(), keys: keys(), values: values(), … }
common.js:218:15
UTTV WARN: No sidebar items found on the current page. Provided nodes: 
NodeList []
​
length: 0
​
<prototype>: NodeListPrototype { item: item(), keys: keys(), values: values(), … }
common.js:218:15
UTTV WARN: No sidebar items found on the current page. Provided nodes: 
NodeList []
​
length: 0
​
<prototype>: NodeListPrototype { item: item(), keys: keys(), values: values(), … }
common.js:218:15
```

Two of the three error messages showed up after I opened up Twitch's front page. The third showed up after I attempted to modify the blacklist with the removal of a single channel. Nothing appeared to pop up in console when I opened it up on the tab featuring the blacklist.

The best I got is what I opened up with in this message, and the reason why I believe that is because a sound editor I sometimes use always opens up with the same settings on a new file, even if I choose different ones, and have to modify the settings _after_ I open up the new file.

What's peculiar to me is that I _can_ modify the blacklist somehow in a couple of ways. From the get-go, I was able to add to the list by clicking on Xs when I synchronized it with the cloud, but I can't remove it. The other way is peculiar, and please pardon the long and layman explanation...

After I mentioned that it was similarly the case on another computer I used, I decided to access that computer again and see if that really was the case, and it's kind of confusing to me that it actually isn't. (Although I am drawing from vague memory; the first time I ever used this program was early-June of last year.)

What I found was, after upgrading my other computer's Firefox to a very similar version to what I'm using on my current computer, I discovered that I could add or subtract categories/channels/tabs _and_ use it without synchronizing it via cloud. I checked to see the installation date and it was early-June of 2019; some weeks before version 19.6.27 came out.

So, I decided to install 19.4.17 on the computer I'm using now to see what would happen, and strangely enough, modifying the list _works_! I could add/subtract from the list of categories, channels, tags, and use it normally without synchronizing to the cloud as well. Obviously, the actual blocking wouldn't work because it's an outdated version, so after modifying the list, I upgraded back to version 20.3.11. The initial problems I described with said version still persists, _but_ the lists I was using when using 19.4.17 are there, and I can only add to the list if I click on Xs. 

So, essentially, if I need to remove something from the list (whether synchronized with the cloud or not), I need to _downgrade_ the program to modify it and then _upgrade_ to use the program with the list. I'm not sure how that came to be, and with a vaguely-similar (but minor) issue with a separate program, I'm left wondering if it's really the program.


My apologies if this wasted your time, sir.

**Comment by kwaschny** [comment]
Thank you for your assistance. The warnings are unrelated and can be ignored.

However, I found a bug that occurs when having non-Twitch tabs open in Firefox. That would also explain why you can add items in twitch.tv tabs, but you cannot modify the blacklist in the management tab. _The dispatcher was assuming the first tab (retrieved through Firefox' tabs API) is a twitch.tv tab and handles the command to save the new/modified blacklist passed from the dispatcher, while the other tabs just syncronize the blacklist in memory. This logic works in Chrome as their tabs API only returns matching tabs for twitch.tv, but fails in Firefox since it retrieves ALL tabs and skips dispatching messages for tabs the extension has no access to, making the save command go into void._

I fixed this in Firefox by just sending the save command to every twitch.tv tab. While this isn't a solid solution, it takes care of the issue for now. I tested it for race conditions and it does not seem to cause trouble. Would you please test the temporary Firefox version `20.3.22` on your machine(s)? @Burg-Eelen It might fix your issue.

To load this temporary version in Firefox, follow these steps:

1. [Download the extension's debug build 20.3.22 here on Github](https://github.com/kwaschny/unwanted-twitch/raw/%2330/publish/firefox/20.3.22.zip).
2. Open Firefox and load the temporary extension on `about:debugging#/runtime/this-firefox` (enter this in your address bar). At the top, there is a button to do so. Choose the downloaded `20.3.22.zip` file and `Unwanted Twitch` should appear as temporary extension. (Temporary extensions are discarded after closing Firefox, so you need to redo this step in case you close Firefox.) It should automatically override the older/installed version of UTTV. You can check the version on `about:addons` (enter this in your address bar).
3. Go to twitch.tv and test whatever you need to test. If you open the developer window in Firefox this time, it will be very chatty and tell you almost everything about what it is doing.

**Comment by lefthand3r** [comment]
for me this fixes the problem (Firefox 74.0 - Mac OSX 10.14.2)

**Comment by kwaschny** [comment]
0c7a52cddf576090120eab9f9e94f02d13605c70 has been published on Firefox Add-ons.
